src/core/common/message.hpp

```diff
@@ -65,6 +63,61 @@ class ThreadLinkInfo;
  *
  */
 
+/**
+ * This macro frees a given message buffer if not nullptr.
+ *
+ * This macro and the ones that follow contain small but common code patterns used in many of the core modules. They
+ * are intentionally defined as macros instead of inline methods/functions to ensure that they are fully inlined.
+ * Note that an `inline` method/function is not necessarily always inlined by the toolchain and not inlining such
+ * small implementations can add a rather large code-size overhead.
+ *
+ * @param[in] aMessage    A pointer to a `Message` to free (can be nullptr).
+ *
+ */
+#define FreeMessage(aMessage)      \
+    do                             \
+    {                              \
+        if ((aMessage) != nullptr) \
+        {                          \
+            (aMessage)->Free();    \
+        }                          \
+    } while (false)
+
+/**
+ * This macro frees a given message buffer if a given `otError` indicates an error.
+ *
+ * The parameter @p aMessage can be nullptr in which case this macro does nothing.
+ *
+ * @param[in] aMessage    A pointer to a `Message` to free (can be nullptr).
+ * @param[in] aError      The `otError` to check.
+ *
+ */
+#define FreeMessageOnError(aMessage, aError)                        \
```

Comment:
What about using static inline function for better type checking. 

```suggestion
static inline void FreeMessageOnError(const Message *&aMessage, otError aError) 
```

Reply:
I used `static inline` first. It can impact code-size (majorly). `inline` is recommendation to the complier. We cannot have such simple method not be inlined (the overhead of calling a function is more than the actual assembly code of the implementation).

Please see here (repeated the same in commit/PR message).
```
/**
 * This macro frees a given message buffer if not nullptr.
 *
 * This macro and the ones that follow contain small but common code patterns used in many of the core modules. They
 * are intentionally defined as macros instead of inline methods/functions to ensure that they are fully inlined.
 * Note that an `inline` method/function is not necessarily always inlined by the toolchain and not inlining such
 * small implementations can add a rather large code-size overhead.
 *
 * @param[in] aMessage    A pointer to a `Message` to free (can be nullptr).
 *
 */```

Reply:
Thanks for the info.